clients/pkg/promtail/targets/cloudflare/target.go

```diff
@@ -0,0 +1,252 @@
+package cloudflare
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/concurrency"
+	"github.com/grafana/dskit/multierror"
+	"github.com/prometheus/common/model"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/positions"
+	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/target"
+
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+var defaultBackoff = backoff.Config{
+	MinBackoff: 1 * time.Second,
+	MaxBackoff: 10 * time.Second,
+	MaxRetries: 5,
+}
+
+type Target struct {
+	logger    log.Logger
+	handler   api.EntryHandler
+	positions positions.Positions
+	config    *scrapeconfig.CloudflareConfig
+	metrics   *Metrics
+
+	client  Client
+	ctx     context.Context
+	cancel  context.CancelFunc
+	wg      sync.WaitGroup
+	to      time.Time // the end of the next pull interval
+	running *atomic.Bool
+	err     error
+}
+
+func NewTarget(
+	metrics *Metrics,
+	logger log.Logger,
+	handler api.EntryHandler,
+	position positions.Positions,
+	config *scrapeconfig.CloudflareConfig,
+) (*Target, error) {
+	if err := validateConfig(config); err != nil {
+		return nil, err
+	}
+	fields, err := Fields(FieldsType(config.FieldsType))
+	if err != nil {
+		return nil, err
+	}
+	client, err := getClient(config.APIToken, config.ZoneID, fields)
+	if err != nil {
+		return nil, err
+	}
+	pos, err := position.Get(positions.CursorKey(config.ZoneID))
+	if err != nil {
+		return nil, err
+	}
+	to := time.Now()
+	if pos != 0 {
+		to = time.Unix(0, pos)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	t := &Target{
+		logger:    logger,
+		handler:   handler,
+		positions: position,
+		config:    config,
+		metrics:   metrics,
+
+		ctx:     ctx,
+		cancel:  cancel,
+		client:  client,
+		to:      to,
+		running: atomic.NewBool(false),
+	}
+	t.start()
+	return t, nil
+}
+
+func (t *Target) start() {
+	t.wg.Add(1)
+	t.running.Store(true)
+	go func() {
+		defer func() {
+			t.wg.Done()
+			t.running.Store(false)
+		}()
+		for t.ctx.Err() == nil {
+			end := t.to
+			maxEnd := time.Now().Add(-time.Minute)
```

Comment:
Should this be configurable?

Reply:
That's in the API definition I can make it const.

Reply:
I think that would help.